### ISSUE TITLE
[IDEA] cascade to provide field-list to tiddlers in both edit and view templates

### ISSUE THREAD

**Issue by AnthonyMuscio** [Jermolene/TiddlyWiki5#6621]
With the introduction of cascades and now the Field Editor cascade we can make much richer field handling. However I may still be unlikely to make use of the Field Editor Cascades, because fields will not be listed in a structured way within the editor. That is the editor is effectively a "poor mans form or dialogue" to edit tiddlers. 

**Describe the solution you'd like**
An additional cascade mechanism or something similar, that can retrieve a different "field-list" for each tiddler according to various conditions and use this to "sortby" in the editor. It could also provide the "field editor(s)" determined by the Field Editor Cascade, to load a field that exists in the field-list, but not in the tiddler (yet).
* Any field that would be displayed normally but not in the field-list will appear at the bottom as it currently does.
* A field named in field-list will be presented for editing even if it is currently hidden from the field editor eg modified.

Ideally if the current tiddler contains a field called field-list it will be used instead of the one found in the cascade mechanism allowing each tiddler to optionally override the field list. Reordered or fields "added/deleted" from the list.

The field list should also be made available to the view template and perhaps later we will have a "field Viewer cascade" as well. ntil then the view template can "iterate" the field-list.

**Describe alternatives you've considered**
I have already considered building my own forms and using the view template, the edit template body or even a custom Story Tiddler cascades. However with the maturation via the Field Editor Cascade I would like this small improvement in the core editor to enable properly managed "Forms" for tiddlers in the core edit template mechanism. Another way is to hide all my custom fields from the field editor and use the $:/tags/EditTemplate to introduce my own form.

**Additional context**
Timely given this bug [[BUG] Field Editor: Switching field name also clears the field value](https://github.com/Jermolene/TiddlyWiki5/issues/6618#)
Allow the rendering of fields within the Edit Template to be extended #6511 

**Example;**
Let say I use the existence of a "contacts" tag or a field "object-type=contacts" to retrieve a field-list for "contacts",

The field-list will be a list of field names perhaps each with a field editor cascade defined for them. 
For Example; `firstname, surname, phone, address1, address2, country, state, postcode, text` 
Note: here I am telling the editor I want these field editor items to be listed before the text field, not below the text field. 

Theoretically this could be used to alter the order of all elements in the editor such as "content type" type and "add new field" could be conditionally displayed or its order set with a virtual fieldname "add-fields".

The editor will use this to list the fields in the editor in the order provided and even the "Field Editor" for fields in the list but not in the tiddler, which are created on providing a value.

**Really simple example**, 
in some very long code tiddlers we may want to see a custom field, or the "add-fields" dialogue displayed above the text field because it is very long to the bottom. field-list would be "custom-fieldname, add-fields, text" and the unnamed type field selector will remain at the bottom.

**Benefits** with the addition of the field-list functionality to the editor mechanism custom forms and field sets could be programmatically applied to provide a robust and usable form handling inside the core features. Field definitions and field-lists could be shared with other wikis to implement common tiddlers with their bundle of fields and field editor cascades.

This will avoid the need for all the different bespoke forms implementations out there (but not take them away) and increase the ability to share "functional tiddler types" such as contacts, customers, books, post, bookmarks, servers, routers, vehicles or any other common "object" etc... found out there.

**Comment by pmario** [comment]
As I wrote at: https://talk.tiddlywiki.org/t/change-physical-order-of-fields-in-a-tiddler/2168/7?u=pmario .. The fields-widget "include" parameter should be able to handle that, if the "sort" parameter is set to no. 

I use the same mechanism to sort my fields in the [config-dialogue](https://wikilabs.github.io/editions/field-search/#%24%3A%2Fplugins%2Fwikilabs%2Ffield-search%2FCP%2FSettings%2Ffield-search) for the field-search plugin. 

The `/sortby` tiddler can contain existing or non existing values. In my usecase "user defined" fields should come first. ..

![image](https://user-images.githubusercontent.com/374655/162949553-37985e5c-6b83-4900-b370-66d2a2122496.png)

**Comment by AnthonyMuscio** [comment]
Ok so it sounds think there is a lot of convergent ideas here. Perhaps too I missed this in the past because there are no examples or code snippets in the FieldsWidget on tiddlywiki.com. Examples of its use in the core are hard to follow eg; $:/core/templates/tid-tiddler.

Using `{{||$:/core/templates/tid-tiddler}}` for example contains some artifacts and the use of the "back tick" eludes me.

However I am keen to see what I propose to be an accessible part of tiddlywiki to all users and hence my suggestion here. Which may after all be provisioned using edit-text and the fields widget, I am keen to see the basic form handling become a high level function.

I will study the fields widget more deeply, if you can show any examples that address the OT more directly I would appreciate it.

**Comment by pmario** [comment]
Also see: https://talk.tiddlywiki.org/t/the-order-of-fields-in-edit-mode/5856/5?u=pmario

>>Mohammad wrote:
>>
>> Would it be possible to include a hidden setting or a configuration tiddler, as mentioned above, to let users customize the order of fields in edit mode without altering the shadow core tiddlers?
>
> Jeremy wrote:
>
>Yes, I’m sure that would be possible. I would be inclined to approach it as a customisable filter that returns the names of fields in the required order. I don’t think it is helpful to make it possible to hide fields in edit mode, and so it would be necessary to also include any fields that are not specified by the filter.

**Comment by pmario** [comment]
Possible, partial solution: https://talk.tiddlywiki.org/t/the-order-of-fields-in-edit-mode/5856/2?u=pmario
